src/main.ts

```diff
@@ -0,0 +1,135 @@
+const repository = process.env.GITHUB_REPOSITORY;
+const branch = process.env.GITHUB_REF;
+const githubEventPayloadFile = process.env.GITHUB_EVENT_PATH || defaultJsonPath;
+const githubEventPayload = require(githubEventPayloadFile);
+const triggerEventName = process.env.GITHUB_EVENT_NAME;
+
+export const parsedNotificationMessage = function parseEventToMessage(eventPayloadText: string): NotificationMessage {
+    let account;
+    let parsedSchema;
+    try {
+        let eventName: string = triggerEventName || 'event';
+        if (core.getInput('event_id')) {
+            eventName = core.getInput('event_id');
+        }
+
+        switch (eventName) {
+            case EventType.PUSH:
+                parsedSchema = schemaOnpush.validateSync(eventPayloadText);
+                account = parsedSchema.pusher.name;
+                return {
+                    message: `**Commit to GitHub** by ${account}`,
+                    url: parsedSchema.compare,
+                    details: `Comment: ${parsedSchema.head_commit.message}`
+                };
+            case EventType.PULL_REQUEST:
+                parsedSchema = schemaOnPullRequest.validateSync(eventPayloadText);
+                account = parsedSchema.pull_request.user.login;
+                return {
+                    message: `**PR submitted to Github** by ${account} with title: ${parsedSchema.pull_request.title}`,
+                    url: parsedSchema.pull_request.html_url,
+                    details: `Pr for merging ref ${parsedSchema.pull_request.head.ref} to base branch ${parsedSchema.pull_request.base.ref}`
+                };
+            case EventType.ISSUE:
+                parsedSchema = schemaOnIssue.validateSync(eventPayloadText);
+                account = parsedSchema.sender.login;
+                return {
+                    message: `**New/updated GitHub issue**: ${parsedSchema.issue.title}`,
+                    url: parsedSchema.issue.html_url,
+                    details: `Issue state: ${parsedSchema.issue.state}  - assignee: ${parsedSchema.issue.assignee}`
+                };
+            case EventType.ISSUE_COMMENT:
+                parsedSchema = schemaOnIssueComment.validateSync(eventPayloadText);
+                return {
+                    message: `**A Github issue comment was posted** by ${account} with content ${parsedSchema.comment.body}`,
+                    url: parsedSchema.issue.html_url,
+                    details: `Issue state: ${parsedSchema.issue.state}  - assignee: ${parsedSchema.issue.assignee}`
+                };
+            default:
+                return {
+                    message: eventName ? eventName.replace(/_/g, '.') : 'A GitHub Actions event has occurred',
+                    url: `https://github.com/${repository}/actions`,
+                    details: core.getInput('details')
+                };
+        }
+    } catch (error) {
+        core.setFailed(error.message);
+        return process.exit(1);
```

Comment:
I don't really understand how this compiles. Seems like `process.exit(1)` is not of type `NotificationMessage`. I think this is quite important because the benefits of types go out of the window quite quickly if the types are a lie.

Reply:
My experience so far : unless I actually returned "anything" in this case, the compiler would complain. Stay tuned, I am testing here with ValidationErrors in my experiment branch, will return with something more "tangible"

Reply:
OK but as it stands you just need to know what is the actual return type of `process.exit`. Let's call it `T`. Once you have that, you know that the return type of the function overall has to be `NotificationMessage | T`.

Reply:
Ok, so some findings : 
- the return is required by the eslint rule https://eslint.org/docs/2.0.0/rules/consistent-return --> the return process.exit(1) is a workaround, so that the linter is satisfied without actually disabling the rule altogether
- the current catch will work for validation errors (the action is failing if validateSync fails) but will not fail if we have an undefined attribute, like pusher.bla  --> however, if the expected attributes are wrong, I expect anyway the validateSync to fail from the beginning so I do not see as an issue 
- my goal was to use the core.setFailed function, which is suggested by github actions. It sets the error code (without us hardcoding it) and fails the actions --> subsequently it returns void. The code for that would be : 
```
export const parsedNotificationMessage = function parseEventToMessage(eventPayloadText: string): NotificationMessage | void {
...
    } catch (error) {
        let errorDetails;
        if (error instanceof ValidationError) {
            errorDetails = error.errors;
        } else {
            errorDetails = error.message;
        }
        return core.setFailed(errorDetails);
    }
...
notifyTeams(<NotificationMessage>parsedNotificationMessage(JSON.stringify(githubEventPayload))); (i have to cast the argument in this case)
```
- another possibility would be to throw specifically an error and also check if it is passed in the notifyTeams functions (based on what I see the flow should never reach there when an error occurs)

Reply:
OK so we have established that the return type of `parseEventToMessage` should be `NotificationMessage | void` which is fine (and sounds correct). 

The only remaining implication of that is that it means that you cannot compose it with a function that expects a NotificatonMessage (because you may not have one). i.e. you cannot pass the result of parseEventMessage directly into notifyTeams without either changing the input arg type of notifyTeams or introducing some code in the middle to check whether you actually have an instance of a NotificationMessage.